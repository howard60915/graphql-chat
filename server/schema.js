const schema = `
input RegisterByEmailMutationInput {
  email: String!
  password: String!
}

type RegisterByEmailMutationPayload {
  status: ResponseStatus
  user: User
}

input UpdateUserMutationInput {
  userId: String!
  email: String
}

type UpdateUserMutationPayload {
  status: ResponseStatus
  user: User
}

input DelUserMutationInput {
  userId: String!
}

type DelUserMutationPayload {
  status: ResponseStatus
  user: User
}

type Mutation {
  registerByEmail(input: RegisterByEmailMutationInput!): RegisterByEmailMutationPayload
  updateUser(input: UpdateUserMutationInput!): UpdateUserMutationPayload
  delUser(input: DelUserMutationInput!): DelUserMutationPayload
}

type Query {
  users: [User]
}

enum ResponseStatus {
  failed
  ok
}

type User {
  id: String
  email: String
  isAdmin: Boolean
}
`;

export default schema;
