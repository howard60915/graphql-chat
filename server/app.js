import graphqlServer from "./graphql";

const PORT = process.env.NODE_PORT || process.env.PORT || 8080;

const options = {
  port: PORT,
  endpoint: "/graphql",
  playground: "/playground"
};

graphqlServer.start(options, ({ port }) => {
  // eslint-disable-next-line no-console
  console.log(`> Ready on http://localhost:${port}`);
});
