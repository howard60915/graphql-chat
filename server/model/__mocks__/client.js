import { Client } from "pg";
import { mock } from "graphql-shortcake";
import knex from "knex";
import config from "../../../knexfile";

const { POSTGRES_HOST } = process.env;

const client = jest.requireActual("../client").default;

const db = knex(config);

const mockClient = mock(client);

mockClient.reset = async name => {
  const pgClient = new Client({
    host: POSTGRES_HOST || "127.0.0.1",
    user: "postgres",
    password: null,
    database: "postgres"
  });
  await pgClient.connect();
  await pgClient.query(`DROP DATABASE IF EXISTS ${name}`);
  await pgClient.query(`CREATE DATABASE ${name}`);
  await pgClient.end();
  await db.migrate.latest();
};

export default mockClient;
