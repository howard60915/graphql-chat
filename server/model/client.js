import { Client } from "graphql-shortcake";
import config from "../../knexfile";

export default new Client({ connectionString: config.connection });
