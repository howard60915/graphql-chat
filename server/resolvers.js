import UsersQuery from "./query/UsersQuery";
import RegisterByEmailMutation from "./mutation/RegisterByEmailMutation";
import UpdateUserMutation from "./mutation/UpdateUserMutation";
import DelUserMutation from "./mutation/DelUserMutation";
import AddMessageMutation from "./mutation/AddMessageMutation";

const Query = {
  users: UsersQuery
};

const Mutation = {
  registerByEmail: RegisterByEmailMutation,
  updateUser: UpdateUserMutation,
  delUser: DelUserMutation,
  addMessage: AddMessageMutation
};

export default { Query, Mutation };
