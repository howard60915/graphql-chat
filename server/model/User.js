import { GlobalModel, Hash } from "graphql-shortcake";
import client from "./client";

export default class User extends GlobalModel {
  static client = client;

  static hasOperator = true;

  static operator = () => User;

  static tableName = "user";

  static columns = {
    email: { type: String },
    password: { type: Hash, paths: ["archive"] },
    isAdmin: { type: Boolean, paths: ["archive"] }
  };

  async checkWritability(user, ThrowError) {
    const result = this.id === user.id || user.isAdmin;
    if (!result) throw new ThrowError();
    return result;
  }
}
