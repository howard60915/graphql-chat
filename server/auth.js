import { NotFoundError } from "graphql-schema-sushi";
import { AuthenticationError } from "apollo-server-errors";
import { JsonWebToken } from "nextjs-apollo/server";
import User from "./model/User";
import Certificate from "./model/Certificate";

const onSignIn = async data => {
  const { email, password, correlationId } = data;

  const { fetcher } = User;
  const user = await fetcher.where("email", email).fetch(NotFoundError);

  if (!(password && user.password.verify(password))) {
    throw new AuthenticationError("AuthenticationError");
  }

  return { user: user.previous, correlationId };
};

const onSignOut = () => ({});

const onRenew = async data => {
  const { user, correlationId } = data;

  const renewUser = await User.load(user.id, NotFoundError);

  return { user: renewUser.previous, correlationId };
};

const toUser = ({ user }) => {
  if (!user.id) throw new AuthenticationError();
  return User.forge(user);
};

const jwt = new JsonWebToken(Certificate, {
  onSignIn,
  onSignOut,
  onRenew,
  toUser
});

export default jwt;
