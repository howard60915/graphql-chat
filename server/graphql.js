import { GraphQLServer } from "graphql-yoga";
import auth from "./auth";
import typeDefs from "./schema";
import resolvers from "./resolvers";

const jwtParser = auth.contextParser();
const server = new GraphQLServer({
  typeDefs,
  resolvers,
  context: async context => {
    const { request, response } = context;
    return jwtParser({ req: request, res: response });
  }
});

auth.applyMiddleware(server.express);

export default server;
