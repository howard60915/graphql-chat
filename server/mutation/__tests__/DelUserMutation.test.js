import client from "../../model/client";
import User from "../../model/User";
import server, { mutation } from "../../__tests__";

const { delUser } = mutation;

describe("DelUserMutation", () => {
  const userId = new User("00000000-0000-0000-0000-000000000010").id;
  const input = { userId };

  afterAll(() => new Promise(() => server.close()));

  it("successfully delete when user is admin", async () => {
    client.mockReturnValueOnce([
      { id: "00000000-0000-0000-0000-000000000010" }
    ]);
    client.mockReturnValueOnce([1]);
    const { errors, data } = await delUser.admin({ input });
    expect(errors).toBeUndefined();
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(2);
    expect(data).toEqual({
      delUser: {
        status: "ok",
        user: { id: userId }
      }
    });
  });

  it("successfully delete when user is self", async () => {
    client.mockReturnValueOnce([
      { id: "00000000-0000-0000-0000-000000000010" }
    ]);
    client.mockReturnValueOnce([1]);
    const { errors, data } = await delUser.user({ input });
    expect(errors).toBeUndefined();
    expect(client).toHaveBeenCalledTimes(2);
    expect(data).toEqual({
      delUser: {
        status: "ok",
        user: { id: userId }
      }
    });
  });

  it("when user is not owner", async () => {
    client.mockReturnValueOnce([
      {
        id: "00000000-0000-0000-0000-000000000015"
      }
    ]);
    const { errors } = await delUser.user({
      input: {
        userId: new User("00000000-0000-0000-0000-000000000015").id
      }
    });
    expect(errors[0].extensions.code).toBe("FORBIDDEN");
  });

  it("when user not found", async () => {
    client.mockReturnValueOnce([]);
    const { errors } = await delUser.user({ input });
    expect(errors[0].message).toBe("not found");
  });
});
