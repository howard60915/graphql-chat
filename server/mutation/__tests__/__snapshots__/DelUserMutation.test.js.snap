// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelUserMutation successfully delete when user is admin 1`] = `
Array [
  "SELECT * FROM \\"user\\" WHERE \\"user\\".\\"deleted_at\\" IS NULL AND \\"id\\" IN ($1)",
  "UPDATE \\"user\\" SET \\"deleted_at\\" = $1, \\"deleted_by\\" = $2 WHERE \\"id\\" = $3 RETURNING *",
]
`;
